### PR TITLE
feat: context-free languages closure reversal

### PR DESCRIPTION
**Pull request by madvorak**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Computability/ContextFreeGrammar.lean [hunk 1]
```diff
@@ -143,3 +146,77 @@ lemma Derives.eq_or_tail {u w : List (Symbol T g.NT)} (huw : g.Derives u w) :
...
+    simp [reverseRule]
+
+private lemma reversed_word_in_original_language {g : ContextFreeGrammar T} {w : List T}
+    (hgw : w ∈ (reverseGrammar g).language) :
+    w.reverse ∈ g.language := by
+  convert derives_reverse hgw
+  simp [List.map_reverse]
+
+/-- The class of context-free languages is closed under reversal. -/
+theorem reverseContextFreeLanguage {L : Language T} (CFL : L.IsContextFree) :
```
Comment (alexjbest) [inline Mathlib/Computability/ContextFreeGrammar.lean:204]:
```suggestion
theorem Language.IsContextFree.reverse {L : Language T} (CFL : L.IsContextFree) :
```

File: Mathlib/Computability/ContextFreeGrammar.lean [hunk 2]
```diff
@@ -143,3 +146,77 @@ lemma Derives.eq_or_tail {u w : List (Symbol T g.NT)} (huw : g.Derives u w) :
...
+private lemma dual_reverseRule {N : Type uN} : reverseRule ∘ @reverseRule T N = id := by
+  ext
+  apply reverseRule_reverseRule
+
+private lemma reverseGrammar_reverseGrammar (g : ContextFreeGrammar T) :
+    reverseGrammar (reverseGrammar g) = g := by
+  cases g with
+  | mk => simp [reverseGrammar, dual_reverseRule, List.map_map, List.map_id]
+
+private lemma derives_reverse {g : ContextFreeGrammar T} {s : List (Symbol T g.NT)}
```
Comment (alexjbest) [inline Mathlib/Computability/ContextFreeGrammar.lean:176]: I'm not sure all of these things should be private, making things private generally just makes it harder for other people to prove things about them

File: Mathlib/Computability/ContextFreeGrammar.lean [hunk 3]
```diff
@@ -143,3 +146,77 @@ lemma Derives.eq_or_tail {u w : List (Symbol T g.NT)} (huw : g.Derives u w) :
   (Relation.ReflTransGen.cases_tail huw).casesOn (Or.inl ∘ Eq.symm) Or.inr
 
 end ContextFreeGrammar
+
+/-- Context-free languages are defined by context-free grammars. -/
+def Language.IsContextFree (L : Language T) : Prop :=
+  ∃ g : ContextFreeGrammar.{uT} T, g.language = L
+
+/-- `L.reverse` is a language that contains exactly all words from `L` backwards. -/
+def Language.reverse (L : Language T) : Language T := { w : List T | w.reverse ∈ L }
```
Comment (alexjbest) [inline Mathlib/Computability/ContextFreeGrammar.lean:155]: should go in Mathlib.Computability.Language

File: Mathlib/Computability/ContextFreeGrammar.lean (hunk 2 above)
Comment (alexjbest) [reply Mathlib/Computability/ContextFreeGrammar.lean:176]: I guess the question is if there are any possible other results whose proof would make use of `reverseGrammar`?

File: Mathlib/Computability/ContextFreeGrammar.lean (hunk 3 above)
Comment (alexjbest) [reply Mathlib/Computability/ContextFreeGrammar.lean:155]: and we should add the basic lemmas for it, i.e. the reverse of the zero and 1 language, reverse of union of two languages, concatenation etc

File: Mathlib/Computability/ContextFreeGrammar.lean (hunk 2 above)
Comment (madvorak) [reply Mathlib/Computability/ContextFreeGrammar.lean:176]: I don't think so.

File: Mathlib/Computability/ContextFreeGrammar.lean [hunk 4]
```diff
@@ -143,3 +146,77 @@ lemma Derives.eq_or_tail {u w : List (Symbol T g.NT)} (huw : g.Derives u w) :
...
+/-- The class of context-free languages is closed under reversal. -/
+theorem Language.IsContextFree.reverse {L : Language T} (CFL : L.IsContextFree) :
+    L.reverse.IsContextFree := by
+  cases CFL with
+  | intro g hgL =>
+    rw [← hgL]
+    use reverseGrammar g
+    apply Set.eq_of_subset_of_subset
+    · intro w hwg
+      exact reversed_word_in_original_language hwg
```
Comment (alexjbest) [inline Mathlib/Computability/ContextFreeGrammar.lean:212]:
```suggestion
    · exact reversed_word_in_original_language
```
Probably you can just write this here

File: Mathlib/Computability/ContextFreeGrammar.lean (hunk 3 above)
Comment (madvorak) [reply Mathlib/Computability/ContextFreeGrammar.lean]: Please check the API.

File: Mathlib/Computability/Language.lean [hunk 5]
```diff
@@ -297,7 +297,7 @@ instance : KleeneAlgebra (Language α) :=
       rw [pow_succ, ←mul_assoc m l (l^n)]
       exact le_trans (le_mul_congr h le_rfl) ih }
 
-/-- Language `l.reverse` is defined as the set of exactly all words from `l` backwards. -/
+/-- Language `l.reverse` is defined as the set of words from `l` backwards. -/
 def reverse (l : Language α) : Language α := { w : List α | w.reverse ∈ l }
 
 lemma reverse_mem_reverse : a.reverse ∈ l.reverse ↔ a ∈ l := by
```
Comment (alexjbest) [inline Mathlib/Computability/Language.lean:303]: simp lemma?

File: Mathlib/Computability/ContextFree.lean [hunk 6]
```diff
@@ -6,14 +6,18 @@ Authors: Martin Dvorak
 import Mathlib.Computability.Language
 
 /-!
-# Context-Free Grammar
+# Context-Free
```
Comment (alexjbest) [inline Mathlib/Computability/ContextFree.lean:9]: what happened to the word grammar, this seems like an odd title now

Comment (madvorak) [reply Mathlib/Computability/ContextFree.lean:9]: I renamed the file. Is it mandatory to change the "docs headline" with it?

Comment (madvorak) [reply Mathlib/Computability/ContextFree.lean]: I should have checked other files. My bad!

**Comment by madvorak** [comment]
I'll split it into two PRs.

**Comment by madvorak** [comment]
https://github.com/leanprover-community/mathlib4/pull/8629

**Comment by madvorak** [comment]
-> https://github.com/leanprover-community/mathlib4/pull/8629

**Pull request closed without merging**